src/main/kotlin/parts/wisdom/arcraider/visualization/Grid.kt

```diff
@@ -1,28 +1,43 @@
 package parts.wisdom.arcraider.visualization
 
 import parts.wisdom.arcraider.SerializedGrid
+import java.awt.Color
+import java.lang.IllegalStateException
 
-data class Coord( val x : Int, val y : Int)
+data class Square(val x : Int, val y : Int, val color: Int)
```

Comment:
Taking this conversation a level higher...

There are 3 types of classes so far:
1) Models that are used to construct the logic side (Grid, Shape, Line, Pixel)
2) Models to build the UI (GridPanel)
3) Glue Models (This for example)

We have a good grasp on the 1st type, even though we still need the Shape implementation and we need to find the limits on how detailed we want to get (do we want Square? Rectangle? Square and Oblong? etc.).
The 2nd type is mostly what you're adding in this review, but the current implementation of Grid blurs the line with the 1st type, since it does kind of both, and the same happens to Square. This 2nd type should just be a single class (or group of classes) that takes a String (or JSON?, whatever) and builds the UI.
The 3rd type in my mind should be just a single Mapper that takes a Grid (1st type) and transforms it into a String or JSON that can be understood by the 2nd group in order to build the UI.

Reply:
I'd lean toward having a set of classes that represent ARC tasks (I'll call these the "ARC classes") as inert data in a "natural", general-purpose way. "Natural" is in the eye of the beholder, but to my eye, a grid would be a basic Kotlin class looking a lot like a matrix of integers, an input/output-grid-pair would be a basic Kotlin class comprising two grids, etc. And by "inert", I mean that although these classes may be mutable, they encompass no logic beyond access and perhaps setting.

I'd lean toward having the UI use the ARC classes directly as its model, rather than transforming them into a more convenient-for-the-UI model. The semantic distance between "natural" ARC classes and a convenient UI model just doesn't seem far enough to me to warrant a separate UI model layer. I could be wrong.

Although the ongoing discussion across the past decade(s) on how best to organize GUIs seems valuable to me, my eyes long since glazed over on views, models, view models, etc. The one dimension of UI design evolution that has held my attention is the criss-crossing paths through ideas like functional reactive programming and frameworks like React that conceptualize the UI as a function from a state to a rendering. But I'm sure you three have great expertise in these areas. Right now we are just trying to do the first simple thing, but perhaps it is worth declaring a UI design pattern that we are following for now, and then bending this work in that direction?

Reply:
I don't feel we're building nearly a complicated enough ui to worry about architectural patterns like MVC etc or maintaining separate models other then where it's useful to separate a MIX class from a non Mix class. If some day we get there can always refactor, but   imo this implementation is fine for now.